torch/distributed/_functional_collectives.py

```diff
@@ -166,6 +166,32 @@ def broadcast(self: torch.Tensor, src: int, group: RANK_TYPES, tag: str = ""):
     return _maybe_wrap_tensor(tensor)
 
 
+def scatter(
+    self: torch.Tensor,
+    scatter_list: Optional[List[torch.Tensor]],
```

Comment:
I want to discuss more on the functional collective API semantic. I am trying to evaluate two options of API:
1. keep this version of API as it matches with the c10d version.
2. we should accept a single tensor (larger tensor), specify a scatter_dim and internally split the big tensor, and return a single tensor (smaller tensor)

@yifuwang which one you think it's better?

Reply:
What's the benefit of splitting a tensor like this? We could add an override for that but we'd still have to split it internally to call the c10d api. I think avoiding api fragmentation would be better if possible (principle of least surprise etc)

Reply:
well, my gut is that the functional APIs should allocate their output buffers and return them. it does feel a bit off to me to have an empty output tensor as input to functional apis.

if we follow that to its conclusion, then you either have to supply tensor_list on every rank, where tensors can be empty/garbage everywhere besides src, but must always have correct shape - or, we could add some 'size' api.

but is it actually better to have a whole tensorlist full of dummy tensors, vs providing an empty output tensor? from the user convenience, i think probably no - maybe its more convenient to provide the empty self tensor.